pkg/apis/seedmanagement/validation/managedseed.go

```diff
@@ -0,0 +1,126 @@
+	allErrs = append(allErrs, ValidateManagedSeedSpec(&managedSeed.Spec, field.NewPath("spec"))...)
+
+	return allErrs
+}
+
+// ValidateManagedSeedUpdate validates a ManagedSeed object before an update.
+func ValidateManagedSeedUpdate(newManagedSeed, oldManagedSeed *seedmanagement.ManagedSeed) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&newManagedSeed.ObjectMeta, &oldManagedSeed.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, ValidateManagedSeedSpecUpdate(&newManagedSeed.Spec, &oldManagedSeed.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateManagedSeed(newManagedSeed)...)
+
+	return allErrs
+}
+
+// ValidateManagedSeedSpec validates the specification of a ManagedSeed object.
+func ValidateManagedSeedSpec(spec *seedmanagement.ManagedSeedSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Check that either seedTemplate or gardenlet is specified
+	if (spec.SeedTemplate == nil && spec.Gardenlet == nil) || (spec.SeedTemplate != nil && spec.Gardenlet != nil) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("seedTemplate/gardenlet"), spec, "exactly one of `seedTemplate` and `gardenlet` is required"))
+	}
+
+	switch {
+	case spec.SeedTemplate != nil:
+		seedTemplatePath := fldPath.Child("seedTemplate")
+
+		// Validate seed template
+		allErrs = append(allErrs, metav1validation.ValidateLabels(spec.SeedTemplate.Labels, seedTemplatePath.Child("metadata", "labels"))...)
+		allErrs = append(allErrs, apivalidation.ValidateAnnotations(spec.SeedTemplate.Annotations, seedTemplatePath.Child("metadata", "annotations"))...)
+		allErrs = append(allErrs, corevalidation.ValidateSeedSpec(&spec.SeedTemplate.Spec, seedTemplatePath.Child("spec"))...)
+
+	case spec.Gardenlet != nil:
+		gardenletPath := fldPath.Child("gardenlet")
+		gardenletConfigPath := gardenletPath.Child("config")
+		seedConfigPath := gardenletConfigPath.Child("seedConfig")
+
+		// Ensure gardenlet config is specified
+		if spec.Gardenlet.Config == nil {
+			allErrs = append(allErrs, field.Required(gardenletConfigPath, "gardenlet config is required"))
+		}
+
+		// Decode gardenlet config from raw extension
+		gardenletConfig, err := helper.DecodeGardenletConfig(spec.Gardenlet.Config)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(gardenletConfigPath, spec.Gardenlet.Config, fmt.Sprintf("could not decode gardenlet config: %v", err)))
+		}
+
+		// Validate gardenlet config
+		allErrs = append(allErrs, configvalidation.ValidateGardenletConfiguration(gardenletConfig)...)
+
+		// Ensure seed config is specified
+		if gardenletConfig.SeedConfig == nil {
+			allErrs = append(allErrs, field.Required(seedConfigPath, "seed config is required"))
+		}
+
+		// Ensure seed name is empty since it will be set by the controller
+		if gardenletConfig.SeedConfig.Name != "" {
```

Comment:
I don't agree. It has to be set by the controller to whatever value it wants (right now it's the managed seed name but we might decide to e.g. use a derivation to ensure uniqueness). Therefore, whatever value we default to won't be used. Therefore, the only sensible check is for an empty value.

Reply:
> It has to be set by the controller to whatever value it wants

Hmm, this is basically defaulting done by a controller. I wouldn't do it this way and rather do it early on during API-level defaulting, which is the proper place for defaulting.

If I remember correctly, we already said, the reference from `ManagedSeed` and the resulting `Seed` object is the name field, though it's not set in stone, yet.
Still, I would actually go with this and only allow `ManagedSeeds` in the `garden` namespace at first (just like `use-as-seed` only works in the `garden` namespace).
This would already solve the uniqueness concern.

If we want to allow creating managed seeds in other namespaces (aka. `PrivateSeeds`) as well later on, we will anyways have to do a lot of other changes, i.e. introduce the `PrivateSeed` resource or something similar.

Reply:
I am still not convinced. I don't think e.g. `Deployment` and other similar resources default the names of the managed `Pods`. I don't see this as "defaulting done by a controller", more as a value that is really owned by the controller, rather than the user. From my perspective, "defaulting" only makes sense if you want to supply a default value for something the user could also supply, but might not care to, which is not the case here.

Note that for consistency, whatever approach we apply to this field should also be applied to `seedTemplate.metadata.name`. `Deployment` and similar resources do not default such names, and they also silently ignore names specified in `template.metadata.name`, without even breaking validation, see https://github.com/kubernetes/kubernetes/blob/6ba09e9a525225430f43233f96fb0c553f6ea1d2/pkg/apis/core/validation/validation.go#L4571 (I also tested it). I would propose a similar approach here. From this perspective even the check above might be seen as redundant, I only added it because it was suggested during previous review rounds.

Reply:
Ok, I'm also not convinced 😄 

I'm more concerned about the reference from ManagedSeed to Seed (both cases with `seedTemplate` and `gardenlet`).
E.g.: how will we find out if a Seed still exists for deletion protection of the `ManagedSeed` if the name is chosen by the controller? We would need to add another field in the status.
I would argue, that for now it's easier to always choose the name of the `ManagedSeed` as the name of the `Seed`. And because we rely on this relationship in the API server, e.g. in admission plugins, we will need to enforce this relationship directly in the API server via defaulting (this is the difference to the deployment controller), because we don't know on which version the reconciling gardenlet runs and what name it is choosing.

So if you think, it will be important in the near feature to have `ManagedSeed.Name != Seed.Name` then let's add a status field. Though, I'm not sure, if we can have an index/fieldselector on a status field, which we would need for efficient lookup.

Reply:
I tend towards @timebertt suggestion here.
When would the managed seed controller need to set another Seed name other than the managed seed resource name?
As pointed out above, the only thing I can think about are uniqueness concerns. 
However, ManagedSeeds that are not created in the `garden` ns should be namespace `PrivateSeeds` I think and therefore not subject to naming conflicts with other Seed clusters.

Reply:
I don't have a strong opinion, but I understand the comparison with `Deployment`/`ReplicaSet` because we are using similar APIs (resource templates that later materialize in a real resource). How do you find out to which `ReplicaSet` a `Pod` belongs or to which `Deployment` a `ReplicaSet` belongs? You can't easily, the only way are the `ownerReferences` (ref #3353). It would have made it much easier if there would be `status` fields. The reason why they do not write the `name` of the to-be-created `ReplicaSet`/`Pod` into the `Deployment`/`ReplicaSet` is because they use `.metadata.generateName` and can't know it upfront, and because there might be multiple resources.

In our case it's a bit easier because there is exactly one `Seed` resource for a `ManagedSeed`. Still, I guess there is no use-case to have different names for `ManagedSeed`s and their owned `Seed`s, is it?

To proceed, I see these three options:

1. We default the seed name to the name of the `ManagedSeed` and validate that both are equal
2. We enforce it's empty and go with the convention that the names are always equal.
3. We enforce it's empty and add a status field for the name that gets populated by the controller when creating the `Seed`.

Personally, I'm fine with all of them, although I see that 3. has some (probably unneeded) overhead. If I had to decide then I'd slightly tend towards 2., simply because it's similar to what I'm used from native Kubernetes resources and because having the name inside the `ManagedSeed` purports that it can be configured while it actually can't.

Reply:
Thanks, I see looking at the problem from different perspectives leads to different solutions. I still tend to vote for option 2 as well.

> how will we find out if a Seed still exists for deletion protection of the ManagedSeed if the name is chosen by the controller? 

I don't fully understand this. We do want to be able to delete a ManagedSeed if its Seed still exists, the controller makes sure that the Seed is properly deleted, similarly to how we today do it when removing the `use-as-seed` annotation. Or do you have something different in mind?

To summarize:
* If there is indeed a case to find the Seed of a ManagedSeed in the API server during admission, etc., then it's an argument for option 1. Though also in this case, we could still list all Seeds and check their owner references to find the right one.
* If there are no such cases, then this strongly supports option 2 IMO.

Reply:
> We do want to be able to delete a ManagedSeed if its Seed still exists, the controller makes sure that the Seed is properly deleted

Yes, you're right. ATM we just protect the `Shoot` from being deleted, when it has the `use-as-seed` annotation and the Seed still exists. But the flow will be different then, you're right.

Having heard all arguments, I also tend to vote for option 2.
So I guess, we're on the same page now, great :)